docs/rfc/0000-application-architecture.md

```diff
@@ -0,0 +1,503 @@
+> configuration, a more robust and flexible solution might be to provide a
+> wrapper command (`rad push/pull`).
+
+The git daemon emits events to the pub-sub bus after successful pushes. Like the
+[post-receive] hook, this is one event per updated ref:
+
+```cddl
+updated = (
+    oid-old: bstr .size 20,
+    oid-new: bstr .size 20,
+    namespace: bstr,
+    ref: tstr        ; relative to namespace
+)
+```
+
+The peer id the git daemon is assuming shall be encoded in the topic name.
+
+
+### CLI
+
+The canonical `radicle-link` CLI shall follow the "subcommand" pattern familiar
+from git and other complex commandline applications. Like git, it shall be
+possible to extend the set of available subcommands by placing executables
+conforming to a naming convention (eg. `rad-<command>`) in the user's PATH. It
+SHALL NOT be possible to override builtin commands by this mechanism.
+
+Each subcommand MUST expose its functionality as a linkable library, and provide
+CBOR serialisation for its arguments and outputs.
+
+To enable [browser applications](#browser-applications), subcommands may be
+callable over the [IPC](#ipc) protocol ("CLIaaS"). The canonical CLI may provide
+a command to bind all available subcommands to a socket wholesale, or a
+configurable subset of them. Unprivileged subcommands may also be exposed as a
+HTTP API (see [Browser Applications](#browser-applications)).
+
+The builtin subcommands shall include network clients for interacting with local
+p2p nodes and the pub-sub bus. The vast majority of new functionality will,
+however, be built on top of the [`Storage`](#storage) primitive. If and when it
+becomes desirable to prototype such functionality in other programming languages
+(ie. without linking to the storage library directly) it will be considered to
+expose the storage interface as a network service itself.
+
+> Note that it does not make much sense to proxy networked subcommands when
+> binding to a socket. Frontends will need to connect to the respective services
+> directly.
+
+A comprehensive specification of the (initial) set of builtins is beyond the
+scope of this document.
+
+
+### Browser Applications
+
+In practice, browser applications are limited to HTTP or WebSockets for network
+calls, filesystem operations must be delegated to a backend service. This poses
+two security concerns:
+
+1. The backend must bind to a well-known TCP port, which is then accessible by
+   all other programs on the same host. Moreover, browsers can be tricked by
+   remote sites into issuing requests against the local host.
+2. Public-key based authentication APIs (such as [WebAuthn][webauthn]) are not
+   yet mature / flexible / portable enough to allow a browser frontend to prove
+   access to the (shared) private key to a backend process.
+
+For these reasons, we devise the following restrictions:
+
+* TCP-based backend services MUST NOT expose privileged operations.
```

Comment:
Does this still hold if we require privileged operations to be signed? (Of course, this doesn’t enable us to build web-based privileged applications yet. It might provide a building block together with browser extensions or WebAuthn though.)

Reply:
No, if we had a way to sign “transactions” in the browser, this would work.

Reply:
I looked into WebAuthn for this a bit, btw, and it makes some ... interesting choices (well, not _that_ surprising given that it’s driven by Yubico to no small extent):

* A “platform authenticator” (TouchID, Windows Hello) is not considered first-class (it does not identify the user), instead, the flows are described such that one first authorises the device using a “roaming authenticator” (USB fob), and only subsequently one may use the platform authenticator. That obviously doesn’t make much sense in a serverless environment.
* Although the spec describes a scenario where a mobile phone is used as a second factor, this doesn’t seem to exist in the wild, let alone in some generic form.
* In general how WebAuthn works is by creating a fresh keypair for each site, which is attested by a hardware key. It does not seem possible to somehow store an existing key with the current browser APIs. I’m not sure what to do with this, given that we’re on the local host: somehow, the hardware key would need to be authorised by the link device key, such that a signature by a key signed by the hardware key is considered eligible. However, the backend still needs to get access to the actual device key!


While I find the idea entertaining that it should only be possible to use radicle link with a hardware token (great for sybil prevention), I get the impression that this is all a bit too convoluted for our purposes.

Reply:
I scrapped the whole idea mostly, just leaving an optional field for request tokens in the IPC protocol. I'm pretty much convinced that we would require an actual second-factor. As a stop-gap, we can expire the key and then prompt for the passphrase every so often.